### DISCUSSION TITLE
v2.0.0

### DISCUSSION THREAD

**Discussion by F0rce** [F0rce/ace#29] [Announcements]
### ⚠ BREAKING CHANGES

* AceEditor extends now from Component

### Features

* added AceReady as Event ([f549839](https://github.com/f0rce/ace/commits/f54983916f11acc4ff3b6e634fa2a3644465cc3e))
* added possibility to add a dynamic autocompletion ([774df2b](https://github.com/f0rce/ace/commits/774df2b498d9aef02a1499a0b04933654f2c4f73)), closes [#13](https://github.com/F0rce/ace/issues/13)
* added working valueChangeListener ([e2a2ef7](https://github.com/f0rce/ace/commits/e2a2ef7908ff11e802a06ca656ab305fea51eefe))
* change style of #editor ([72795b3](https://github.com/f0rce/ace/commits/72795b3071e99f98c72607974db18a893aafe7f2))
* convenience methods ([9b6a23d](https://github.com/f0rce/ace/commits/9b6a23d8656a271cd6725faeef14c9307fe5495f))
* **modes:** adding lobster_logs & lobster_records ([d651b3d](https://github.com/f0rce/ace/commits/d651b3df4422333558e5fac2a3b0a4862767b116))
* using new AceSelection and AceCursorPosition Classes ([40de358](https://github.com/f0rce/ace/commits/40de3581af32739d457e24df3b01b98e3115117e))

### Bug Fixes

* AceEditor extends now from Component ([a98c1d3](https://github.com/f0rce/ace/commits/a98c1d360d9065f6b56108fb6cf52cbaf9891916))
* removing close to all racing conditions ([774df2b](https://github.com/f0rce/ace/commits/774df2b498d9aef02a1499a0b04933654f2c4f73))

<hr /><em>This discussion was created from the release <a href='https://github.com/F0rce/ace/releases/tag/v2.0.0'>v2.0.0</a>.</em>

**Comment by F0rce** [comment]
Hello everyone,

this is one of the biggest releases ace has seen since the beginning of the project. If you miss any functionality or have some questions be sure to drop them in here. 

I tested this release **a lot**, but maybe in some cases (which I didn't cover) there could be some problems *(I really hope not)*.

Huge thanks to @sirbris for testing everything with me and for his large influence on how this release turned out (speaking features and most importantly bug fixes). 

Have a great week,
David

**Comment by F0rce** [comment]
Patch [2.0.1](https://github.com/F0rce/ace/releases/tag/v2.0.1) was just published.

When you dettached ace (Vaadin removes it from the dom) and reattached the same instance of the editor, the value was lost. As some people may do that, I thought it would be important to patch it immediately.

For any further questions, be sure to drop them in here ;)

**Comment by F0rce** [comment]
Minor [2.1.0](https://github.com/F0rce/ace/releases/tag/v2.1.0) was just published.

For any further questions, be sure to drop them in here ;)

**Comment by askpythia** [comment]
Hi David,
Is there a reason why V2.x ff does not implement the Vaadin HasValue interface anymore?

**Reply by F0rce** [reply]
Hei @askpythia,

as I was overwriting all of the methods implemented with the HasValue interface, I decided to remove it. I thought there was no need to keep a "unused" interface. Did I break some of your logic ?

**Reply by askpythia** [reply]
In my case I've some Methods which have a HasValue as Parameter, so I've recognized the API break. With this I can handle different Widgets the same way...

Vaadin has a very good design using all the Has... Interfaces to allow a very generic and modular design, would be nice if Ace - as long as it is a Widget with a (String) value would implement HasValue, too.

**Reply by F0rce** [reply]
I'll give it some consideration and maybe include it in a future minor / patch version. There are still some things I want to optimize regarding Ace, so I think a new version should not be that far from now. I'll keep you updated.
